server/schedule/placement/rule_manager.go

```diff
@@ -275,10 +276,51 @@ func (m *RuleManager) FitRegion(stores StoreSet, region *core.RegionInfo) *Regio
 	return FitRegion(stores, region, rules)
 }
 
-func (m *RuleManager) swapRule(rule *Rule) *Rule {
+func (m *RuleManager) tryBuildSave(oldRules map[[2]string]*Rule) error {
+	ruleList, err := buildRuleList(m.rules)
+	if err == nil {
+		for key := range oldRules {
+			rule := m.rules[key]
+			if rule != nil {
+				err = m.store.SaveRule(rule.StoreKey(), rule)
+			} else {
+				r := Rule{
+					GroupID: key[0],
+					ID:      key[1],
+				}
+				err = m.store.DeleteRule(r.StoreKey())
+			}
+			if err != nil {
+				// TODO: it is not completely safe
+				// 1. in case that half of rules applied, error.. we have to cancel persisted rules
+				// but that may fail too, causing memory/disk inconsistency
```

Comment:
It is admirable for you to notice this issue! I think we can add some note in the API annotation.